src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2581,6 +2582,36 @@ OpenNewConnections(WorkerPool *workerPool, int newConnectionCount,
 			connectionFlags |= OUTSIDE_TRANSACTION;
 		}
 
+		/*
+		 * Before opening the second connection to a worker, it is time to mark
+		 * the parallel execution. This is useful for supporting foreign keys to
+		 * reference tables. See details in
+		 * EnforceRestrictionsOnParallelRelationAccess().
+		 */
+		if (list_length(workerPool->sessionList) == 1)
+		{
+			DistributedExecution *execution = workerPool->distributedExecution;
+
+			/*
+			 * Recording parallel relation access might trigger the
+			 * execution to be finished sequentially.
+			 */
+			EnforceRestrictionsOnParallelRelationAccess(execution->tasksToExecute);
+			if (MultiShardConnectionType == SEQUENTIAL_CONNECTION)
+			{
+				execution->targetPoolSize = 1;
```

Comment:
I am a bit confused by this. We're subtly changing two important global states in the middle of a function called "OpenNewConnections". We should at least move it out of OpenNewConnections.

It's also a bit weird to me to retroactively change the target pool size. Could we have a situation where we already have 2 connections, but later change the pool size to 1?

Reply:
I suggest that we do such important changes over methods to abstract the process. For example if we used a method to change the targetPoolSize then later we can add a new constraint or we can have extra checks in the method. What if we cannot set it to 1 when it is a certain value in the future or even now? For someone without context, it might not be clear what `setting targetPoolSize to 1` means, however if we had a method like:

```c
void UseSingleConnectionPerNode(execution) {
     if (some condition) {
          ereport ( 'cannot switch to a single connection because of ....');
     }
     execution->targetPoolSize = 1;
}
```

What do you think of this?

Reply:
We'll likely to skip this PR due to this: https://github.com/citusdata/citus/pull/4141#issuecomment-688172555